subprocess/remote_command.go

```diff
@@ -46,28 +47,51 @@ func (rc *RemoteCommand) Run(ctx context.Context) error {
 		grip.Warningf("RemoteCommand(%s) has nil Cmd or Cmd.Process in Run()", rc.Id)
 	}
 
+	chckCtx, cancel := context.WithCancel(ctx)
+
 	errChan := make(chan error)
+
 	go func() {
 		errChan <- rc.Cmd.Wait()
 	}()
 
+	go func() {
+		timer := time.NewTimer(0)
+		defer timer.Stop()
+		startAt := time.Now()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				if rc.Cmd.ProcessState.Exited() {
+					grip.Info(message.Fields{
+						"message": "remote command process ended early",
+						"id":      rc.Id,
+						"cmd":     rc.CmdString,
+						"runtime": time.Since(startAt),
+						"span":    time.Since(startAt).String(),
+					})
+					cancel()
+					return
+				}
+				timer.Reset(100 * time.Millisecond)
+			}
+		}
+	}()
+
 	select {
-	case <-ctx.Done():
-		err = rc.Cmd.Process.Kill()
-		return errors.Wrapf(err,
-			"operation '%s' was canceled and terminated.",
-			rc.CmdString)
 	case err = <-errChan:
 		return errors.WithStack(err)
+	case <-chckCtx.Done():
+		return nil
+	case <-ctx.Done():
```

Comment:
if `ctx` is canceled, `chckCtx` is also canceled, which makes this non-deterministic.

Reply:
I made a separate context for this